### ISSUE TITLE
Intermittent test failure: truncated lock file

### ISSUE THREAD

**Issue by jkeiser** [chef/chef#3961]
[This test](https://github.com/chef/chef/blob/master/spec/functional/run_lock_spec.rb#L280) failed last night on master on OS/X:

```
 1) Chef::RunLock when locking the chef-client run doesn't truncate the lock file so that contents can be read
     Failure/Error: expect(File.read(lockfile)).to eq(p1.to_s)
       
       expected: "37461"
            got: ""
       
       (compared using ==)
     # ./spec/functional/run_lock_spec.rb:280:in `block (3 levels) in <top (required)>'
```

**Comment by jkeiser** [comment]
Placing in the 12.5 milestone until we know it's not a regression.

**Comment by jkeiser** [comment]
It looks like there is a race condition in our runlock reader code.  There is a heartbeat between when the lockfile is created (empty) and when we acquire the lock to write it, since the `File.open` and `flock` happen at separate times.

The most obvious thing to do would be to open the files already locked, and have the reader do a lock. Some Unix systems allow you to pass `LOCK_EX` to `fopen`, so this is at least possible. But it's unlikely that all systems support this, and besides, having the readers also do `LOCK_EX` will increase contention and possibly send false signals that a process is actually running (since we use the existence of the lock as a part of that signal).

The solution seems like to spin the reader if the file is empty, up to a timeout.

**Comment by jkeiser** [comment]
Fixed by #3967.
